crates/wasmparser/src/validator/component.rs

```diff
@@ -2266,62 +2369,23 @@ impl ComponentState {
             }
         }
 
-        // Validate various combinations of options with respect to async.
-        // Modeled as a `match` here to double-check that everything is
-        // exhaustive at compile-time.
-        match (
-            async_,
-            allow_async,
-            callback.is_some(),
-            core_ty,
-            post_return.is_some(),
-        ) {
-            (true, false, ..) => bail!(offset, "async option not allowed here"),
-            (false, _, true, _, _) => {
-                bail!(offset, "cannot specify callback without lifting async")
-            }
-            (true, true, _, _, true) => {
-                bail!(
-                    offset,
-                    "cannot specify post-return function when lifting async"
-                )
+        let concurrency = match (is_async, callback, post_return.is_some()) {
+            (false, Some(_), _) => {
+                bail!(offset, "cannot specify callback without async")
             }
-
-            // Async + allowed + this is a lift (core_ty present) + stackful ABI
-            (true, true, false, Some(_), false) => {
-                if !self.features.cm_async_stackful() {
-                    bail!(
-                        offset,
-                        "`async` without `callback` requires the async stackful feature"
-                    )
-                }
+            (true, _, true) => {
+                bail!(offset, "cannot specify post-return function in async")
             }
+            (false, _, _) => Concurrency::Sync,
```

Comment:
Can this bind `false, None, _` to guarantee the callback isn't provided at-a-glance?